server/server/src/billing/billing_service.rs

```diff
@@ -1,194 +1,361 @@
+                &info.subscription_product_id,
+                &info.purchase_token,
+            ).await?;
 
-    let tx_result = tx!(con, pipe, &[stripe_user_info(public_key)], {
-        user_info = con
-            .maybe_json_get(stripe_user_info(public_key))
-            .await?
-            .unwrap_or_default();
+            info.account_state = GooglePlayAccountState::Canceled;
+            info!("Successfully canceled google play subscription of user. public_key: {:?}.", context.public_key);
 
-        let current_time = get_time().0 as Timestamp;
+            // You do not get rid of the data cap until they have used up the rest of their time.
+        }
+        Some(BillingPlatform::Stripe(ref info)) => {
+            info!("Canceling stripe subscription of user. public_key: {:?}.", context.public_key);
+
+            stripe_client::cancel_subscription(
+                &server_state.stripe_client,
+                &stripe::SubscriptionId::from_str(&info.subscription_id)?,
+            )
+                .await
+                .map_err(|err| internal!("{:?}", err))?;
+
+            sub_profile.data_cap = FREE_TIER_USAGE_SIZE;
+            info!("Successfully canceled stripe subscription. public_key: {:?}", context.public_key);
 
-        if current_time - user_info.last_in_payment_flow < millis_between_payment_flows {
-            info!(
-                "User is already in payment flow, or this request is too soon after a failed one. public_key: {}",
-                keys::stringify_public_key(public_key)
-            );
-            return Err(Abort(ClientError(SwitchAccountTierError::ConcurrentRequestsAreTooSoon)));
         }
+    }
+
+    reset_billing_lock(&context.public_key, &mut con, &mut sub_profile).await?;
 
-        user_info.last_in_payment_flow = current_time;
+    Ok(CancelSubscriptionResponse {})
+}
+
+pub async fn upgrade_account_stripe(
+    context: RequestContext<'_, UpgradeAccountStripeRequest>,
+) -> Result<UpgradeAccountStripeResponse, ServerError<UpgradeAccountStripeError>> {
+    let (request, server_state) = (&context.request, context.server_state);
 
-        pipe.json_set(stripe_user_info(public_key), &user_info)?;
+    info!("Attempting to switch account tier of {:?} to premium", context.public_key);
 
-        Ok(&mut pipe)
+    let mut con = server_state.index_db_pool.get().await?;
+
+    let mut sub_profile = lock_billing_workflow(
+        &context.public_key,
+        &mut con,
+        server_state.config.stripe.millis_between_user_payment_flows,
+    )
+    .await
+    .map_err(|err| match err {
+        ClientError(LockBillingWorkflowError::ExistingRequestPending) => {
+            ClientError(UpgradeAccountStripeError::ExistingRequestPending)
+        }
+        InternalError(msg) => InternalError(msg),
+    })?;
+
+    if sub_profile.data_cap == PREMIUM_TIER_USAGE_SIZE {
+        return Err(ClientError(UpgradeAccountStripeError::AlreadyPremium));
+    }
```

Comment:
Does this mean that no one can switch between android -> strip or reverse without cancelling? But also maybe they're just updating card info (they plan to cancel a card), wouldn't we prevent them from doing that?

I think we need a doc that rfc's these interactions a bit. I think you'd want to include a drawing of a finite-state-machine that shows what various states you can go to (or cannot) what info is needed (or why you cannot go between those states).

I think we should build on that doc to determine what our account cancellation / declines / etc strategy is.

You need to capture what we have now in this PR and what target state may look like, and we can determine then what this PR needs now v. later.

I can do a good bit of this when it's time for me to implement apple, but start the doc now because you know the details of 2/3 of the payment processors we plan to support immediately.